### PR TITLE
Enh/default cri

### LINKED ISSUE #1038: Make `docker` container runtime a first-class citizen
**What would you like to be added**:
* Read list of supported runtimes from `CloudProfile` only, don't add `docker` to the list of available container runtimes for a `MachineImageVersion`
* When `docker` is selected, set `cri.name=docker` explicitly, don't leave `cri=nil`
* This concerns both aspects: adding a new cluster and/or worker pool as well as modifying an existing worker pool
* Notice to the release notes that open-source users have to adapt their `CloudProfile` to explicitly add `docker` to all `MachineImageVersions` which support it.

**Why is this needed**:
In the context of deprecating the in-tree dockershim, we're making a few changes to g/g, which will result in shoots with k8s version >=1.22 getting `containerd` as container runtime when `cri=nil`:
* As a user, I want to be able to explicitly configure docker as a container runtime, such that I can switch back to docker when I find out my workload doesn't run nicely on containerd: gardener/gardener#4218
* As a gardener operator, I want new shoots to be created with containerd as a default container runtime, such that users no longer get shoots using deprecated functionality: gardener/gardener#4222

**Note**
Prerequisites for this change are 
* As a user, I don't want a node rollout to happen when my configuration change is essentially a no-op, such that I don't suffer from unnecessary downtime (cases: explicitly specifying docker when previously created with docker as implicit default): gardener/gardener#4237
* Adapting the `CloudProfile` as mentioned above before deploying

This is  a prerequisite for releasing a g/g version which supports k8s 1.22 (earliest in August).

cc @voelzmo

### LINKED ISSUE #1039: Select `containerd` as default for cri.Name in shoots with k8s version >= 1.22
**What would you like to be added**:
When creating a new cluster/worker in a shoot with k8sVersion >= 1.22 then preselect `containerd` as default `Container Runtime`. Currently, `docker` is preselected:
image

**Why is this needed**:
In g/g with k8sversion >= 1.22 we default the containerRuntime to `containerd` if `cri == nil`. The intention is to keep the dashboard experience consistent.
Background:
* As a gardener operator, I want new shoots to be created with containerd as a default container runtime, such that users no longer get shoots using deprecated functionality: gardener/gardener#4222

**Note**
It is possible that the user selects a Machine Image which does not support `containerd`. 

cc @voelzmo

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
- Select containerd as default for cri.Name in shoots with k8s version >= 1.22
- Make docker container runtime a first-class citizen 

**Which issue(s) this PR fixes**:
Fixes #1039 Fixes #1038

**Special notes for your reviewer**:

**Release note**:

```breaking operator
The Dashboard no longer adds `docker`to the list of available CRIs. You need to adapt all `CloudProfiles` and explicitly add `docker` to all MachineImageVersions which support it
```
```feature user
`Container Runtime` is now a required field for cluster workers and defaulted to `containerd` for cluster kubernetes versions 1.22 and higher. Clusters with older kubernetes versions keep `docker` as default container runtime. If default runtime is not in the list of supported runtimes of a machine image it defaults to the first one specified in the cloud profile
```

### CODE REVIEW

File: frontend/src/utils/index.js [hunk 1]
```diff
@@ -504,6 +504,13 @@ export function maintenanceWindowWithBeginAndTimezone (beginTime, beginTimezone,
   return { begin, end }
 }
 
+export function defaultCRIForWorker (kubernetesVersion, containerRuntimes) {
+  if (semver.lt(kubernetesVersion, '1.22.0')) {
+    return includes(containerRuntimes, 'docker') ? 'docker' : head(containerRuntimes)
+  } else {
+    return includes(containerRuntimes, 'containerd') ? 'containerd' : head(containerRuntimes)
+  }
+}
```
Comment (holgerkoser) [inline frontend/src/utils/index.js:513]:
```suggestion
export function defaultCRIForWorker (kubernetesVersion, containerRuntimes) {
  const containerRuntime = semver.lt(kubernetesVersion, '1.22.0') 
    ? 'docker' 
    : 'containerd'
  return includes(containerRuntimes, containerRuntime) 
    ? containerRuntime 
    : head(containerRuntimes)
}
```

File: frontend/src/store/modules/shoots/index.js [hunk 2]
```diff
@@ -228,7 +228,7 @@ const actions = {
       set(shootResource, 'spec.provider.infrastructureConfig.networks.zones', zonesNetworkConfiguration)
     }
 
-    const worker = omit(rootGetters.generateWorker(zones, cloudProfileName, region), ['id', 'isNew'])
+    const worker = omit(rootGetters.generateWorker(zones, cloudProfileName, region, kubernetesVersion.version), ['id', 'isNew'])
```
Comment (holgerkoser) [inline frontend/src/store/modules/shoots/index.js:231]:
```suggestion
    const newWorker = rootGetters.generateWorker(zones, cloudProfileName, region, kubernetesVersion.version)
    const worker = omit(newWorker, ['id', 'isNew'])
```

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 3]
```diff
@@ -70,50 +64,63 @@ export default {
...
     containerRuntimeItems () {
-      const containerRuntimes = map(this.machineImageCri, 'name')
-      return uniq([...containerRuntimes, DEFAULT_CONTAINER_RUNTIME])
+      return map(this.machineImageCri, 'name')
     },
     ociRuntimeItems () {
       const containerRuntime = find(this.machineImageCri, ['name', this.containerRuntime])
-      const ociRuntimes = get(containerRuntime, 'containerRuntimes', [])
-      return map(ociRuntimes, 'type')
+      const ociRuntimess = get(containerRuntime, 'containerRuntimes', [])
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:86]: I am still confused by the naming

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 4]
```diff
@@ -70,50 +64,63 @@ export default {
...
+      return map(ociRuntimess, 'type')
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:87]:
```suggestion
      const ociRuntimes = get(containerRuntime, 'containerRuntimes', [])
      return map(ociRuntimes, 'type')
```

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 5]
```diff
@@ -70,50 +64,63 @@ export default {
...
+    },
+    containerRuntime: {
+      get () {
+        return get(this.worker, 'cri.name')
+      },
+      set (value) {
+        set(this.worker, 'cri.name', value)
+      }
+    },
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:96]: Why do we use the name `containerRuntime` for the `cri.name`.

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 6]
```diff
@@ -70,50 +64,63 @@ export default {
...
-      ociRuntime: undefined
+      valid: undefined
     }
   },
   validations,
   computed: {
     containerRuntimeItems () {
-      const containerRuntimes = map(this.machineImageCri, 'name')
-      return uniq([...containerRuntimes, DEFAULT_CONTAINER_RUNTIME])
+      return map(this.machineImageCri, 'name')
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:82]: I would propose to rename this property to `criNames`

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue (hunk 3 above)
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:86]: I  would also rename `ociRuntimeItems` to `criContainerRuntimeTypes`. This helps me to map the property to the corresponding part in the cloudProfile.

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 7]
```diff
@@ -70,50 +64,63 @@ export default {
...
     },
     ociRuntimeItems () {
       const containerRuntime = find(this.machineImageCri, ['name', this.containerRuntime])
-      const ociRuntimes = get(containerRuntime, 'containerRuntimes', [])
-      return map(ociRuntimes, 'type')
+      const ociRuntimess = get(containerRuntime, 'containerRuntimes', [])
+      return map(ociRuntimess, 'type')
+    },
+    containerRuntime: {
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:89]: Rename `containerRuntime` to `criName`

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 8]
```diff
@@ -70,50 +64,63 @@ export default {
...
+      get () {
+        return get(this.worker, 'cri.name')
+      },
+      set (value) {
+        set(this.worker, 'cri.name', value)
+      }
+    },
+    ociRuntimes: {
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:97]: rename `ociRuntimes` to `selectedCriContainerRuntimeTypes`

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 9]
```diff
@@ -70,50 +64,63 @@ export default {
...
+        return map(ociRuntimes, 'type')
+      },
+      set (value) {
+        if (value && value.length) {
+          const containerRuntimes = map(value, ociRuntimes => ({ type: ociRuntimes }))
+          set(this.worker, 'cri.containerRuntimes', containerRuntimes)
+        } else {
+          unset(this.worker, 'cri.containerRuntimes')
+        }
+      }
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:112]:
```suggestion
      get () {
        const criContainerRuntimes = get(this.worker, 'cri.containerRuntimes')
        return criContainerRuntimes 
          ? map(criContainerRuntimes, 'type')
          : undefined
      },
      set (value) {
        if (!isEmpty(value)) {
          const criContainerRuntimes = map(value, type => ({ type }))
          set(this.worker, 'cri.containerRuntimes', criContainerRuntimes)
        } else {
          unset(this.worker, 'cri.containerRuntimes')
        }
      }
```

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue (hunk 6 above)
Comment (grolu) [reply frontend/src/components/ShootWorkers/ContainerRuntime.vue:82]: ok, done

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue (hunk 3 above)
Comment (grolu) [reply frontend/src/components/ShootWorkers/ContainerRuntime.vue]: ok, done

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue (hunk 5 above)
Comment (grolu) [reply frontend/src/components/ShootWorkers/ContainerRuntime.vue:96]: ok, done

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue (hunk 7 above)
Comment (grolu) [reply frontend/src/components/ShootWorkers/ContainerRuntime.vue:89]: ok, done

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue (hunk 8 above)
Comment (grolu) [reply frontend/src/components/ShootWorkers/ContainerRuntime.vue:97]: ok, done

File: frontend/src/components/ShootWorkers/ContainerRuntime.vue [hunk 10]
```diff
@@ -122,13 +129,16 @@ export default {
...
-    if (ociRuntime) {
-      this.ociRuntime = map(ociRuntime, 'type')
-    }
     this.$v.$touch()
     this.validateInput()
+  },
+  watch: {
+    criNames (criNames) {
+      if (!includes(criNames, this.criName)) {
+        this.criName = defaultCriNameByKubernetesVersion(this.criNames, this.kubernetesVersion)
```
Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:138]:
```suggestion
    criNames (value) {
      if (!includes(value, this.criName)) {
        this.criName = defaultCriNameByKubernetesVersion(value, this.kubernetesVersion)
```

Comment (holgerkoser) [inline frontend/src/components/ShootWorkers/ContainerRuntime.vue:138]:
```suggestion
        this.criName = defaultCriNameByKubernetesVersion(criNames, this.kubernetesVersion)
```

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**